File: cpp/DBBD/DBBD/TcpClientSession.cpp
#include "TcpClientSession.h"

DBBD::TcpClientSession::TcpClientSession(IoContextSP context, SocketSP socket, unsigned int bufferSize,
	std::function<void(DBBD::Buffer)> method)
	: TcpSessionBase(context, socket, bufferSize)
	, bindReadInternal(method)
{
}

DBBD::TcpClientSession::~TcpClientSession()
{
}

void DBBD::TcpClientSession::startInternal()
{
}

void DBBD::TcpClientSession::readInternal(DBBD::Header header)
{
	char* block = readBuffer->readByteBlock(header.length);
	Buffer buffer(block, header.length);
	bindReadInternal(buffer);
}

void DBBD::TcpClientSession::writeInternal()
{
}

File: csharp/DBBD/DBBD/Network/TcpClientSession.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Text;

namespace DBBD
{
    public class TcpClientSession : TcpSessionBase
    {
        public TcpClientSession(Socket socket, uint bufferSize, Action<Buffer> method)
            : base(socket, bufferSize)
        {
            this.bindReadInternal = method;
        }

        protected override void ReadInternal(Header header)
        {
            byte[] block = readBuffer.ReadByteBlock(header.Length);
            Buffer buffer = new Buffer(block);
            bindReadInternal(buffer);
        }

        protected override void StartInternal()
        {
        }

        protected override void WriteInternal()
        {
        }

        private Action<Buffer> bindReadInternal;
    }
}
